BUG: Fix weird behavior with mask and units (Fixes #8908)

dopplershift opened the pull request:
Refactor many places that check for masked arrays, and fill the masked
values with nans, with a helper to accomplish that. In the helper,
replace the isinstance check with a attribute check for 'mask'. This
allows libraries, like pint, that wrap mask arrays to pass the check and
be converted appropriately.

Also fix one spot using `atleast_1d` with `_check_1d`.

tacaswell commented on lib/matplotlib/cbook/__init__.py:
```diff
@@ -1963,6 +1963,17 @@ def is_math_text(s):
+    if hasattr(x, 'mask'):
+        return np.ma.asarray(x, float).filled(np.nan)
+    else:
+        return np.asarray(x, float)
```
is `asanyarray` better here?

dopplershift replied:
The existing code was always going to a straight-up numpy array, and since that part of the behavior didn't seem to be causing any unit-related problems, it seemed best to be conservative and keep it as `asarray`.

dopplershift commented:
Looks like I already started a small pint knock-off in `test_units.py`. I'll add a test in there.

dopplershift commented:
Added a test. Also rolled back changes to `collections.py`; these were adding a conversion to `array` that isn't appropriate given that things like `set_verts` take lists of arrays--causing broken tests.

dopplershift commented:
IMO, this is good to go.

dopplershift commented:
I went ahead and added the test I developed for #8910. The only outstanding question on that is whether the call to `autoscale_view()` should really be necessary.

dopplershift commented:
Went ahead and added a new callback for the `'units finalize'` event that calls both `relim` and `autoscale_view` (was just `relim` before). Also rebased on lastest master and fixed a pep8 violation.

dopplershift commented:
Fixed tests by making sure to only autoscale relevant axis.

dopplershift commented:
Ok, Travis passed, but for some reason isn't reporting.

tacaswell commented:
I suspect that this will make @TD22057 happy.
